File: 20. NetworkDelayTime.py
"""
Task: Network Delay Time
Given a network of n nodes, and a list of travel times as directed edges. If a signal is sent from a source node, find how long it will take for all nodes to receive the signal.
Application of Dijkstra's Algorithm.
"""
import heapq
def network_delay_time(times, n, k):
    adj = [[] for _ in range(n+1)]
    for u,v,w in times:
        adj[u].append((v,w))
    dist = [float('inf')] * (n+1)
    dist[k] = 0
    heap = [(0, k)]
    while heap:
        d, u = heapq.heappop(heap)
        if d > dist[u]: continue
        for v, w in adj[u]:
            if dist[v] > dist[u] + w:
                dist[v] = dist[u] + w
                heapq.heappush(heap, (dist[v], v))
    ans = 0
    for i in range(1, n+1):
        if dist[i] == float('inf'):
            return -1
        ans = max(ans, dist[i])
    return ans
# Sample Test Case
if __name__ == "__main__":
    times = [[2,1,1],[2,3,1],[3,4,1]]
    n, k = 4, 2
    print("Network Delay Time:", network_delay_time(times, n, k))
"""
Time Complexity: O((V+E)logV)
Space Complexity: O(V+E)
"""

File: 20. NetworkDelayTime.cpp
/*
Task: Network Delay Time
Given a network of n nodes, and a list of travel times as directed edges. If a signal is sent from a
source node, find how long it will take for all nodes to receive the signal. Application of
Dijkstra's Algorithm.
*/
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

int networkDelayTime(vector<vector<int>>& times, int n, int k)
{
    vector<vector<pair<int, int>>> adj(n + 1);
    for (auto& t : times) adj[t[0]].push_back({t[1], t[2]});
    vector<int> dist(n + 1, INT_MAX);
    dist[k] = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, k});
    while (!pq.empty())
    {
        int d = pq.top().first, u = pq.top().second;
        pq.pop();
        if (d > dist[u])
            continue;
        for (auto& edge : adj[u])
        {
            int v = edge.first, w = edge.second;
            if (dist[v] > dist[u] + w)
            {
                dist[v] = dist[u] + w;
                pq.push({dist[v], v});
            }
        }
    }
    int ans = 0;
    for (int i = 1; i <= n; ++i)
        if (dist[i] == INT_MAX)
            return -1;
        else
            ans = max(ans, dist[i]);
    return ans;
}
// Sample Test Case
int main()
{
    vector<vector<int>> times = {{2, 1, 1}, {2, 3, 1}, {3, 4, 1}};
    int n = 4, k = 2;
    cout << "Network Delay Time: " << networkDelayTime(times, n, k) << endl;
    return 0;
}
/*
Time Complexity: O((V+E)logV)
Space Complexity: O(V+E)
*/
